crates/shared/src/bad_token/trace_call.rs

```diff
@@ -61,59 +67,68 @@ impl TraceCallDetector {
+    }
+
     // Err if communication with the node failed.
     // Ok(None) if there is no pool or getting the balance fails.
     // Ok(address, balance) for an address that has this amount of balance of the token.
-    async fn find_largest_pool_owning_token(&self, token: H160) -> Result<Option<(H160, U256)>> {
-        let mut candidates = HashSet::new();
-        for result in futures::future::join_all(
-            self.finders
-                .iter()
-                .map(|finder| finder.find_candidate_owners(token)),
-        )
-        .await
-        {
-            candidates.extend(match result {
-                Ok(candidates) => candidates,
-                Err(err) => {
-                    tracing::warn!("token owner finding failed: {:?}", err);
-                    continue;
-                }
-            });
-        }
-
+    async fn find_pool_owning_token(
+        &self,
+        token: H160,
+        min_amount: U256,
+    ) -> Result<Option<(H160, U256)>> {
         const BATCH_SIZE: usize = 100;
         let instance = ERC20::at(&self.web3, token);
-        let mut batch = CallBatch::new(self.web3.transport());
-        let futures = candidates
-            .iter()
-            .map(|&address| {
-                let fut = instance.balance_of(address).batch_call(&mut batch);
-                async move { (address, fut.await) }
-            })
-            .collect::<Vec<_>>();
-        batch.execute_all(BATCH_SIZE).await;
-
-        let mut biggest_balance = None;
-        for future in futures {
-            let (address, result) = future.await;
-            let balance = match result {
-                Ok(balance) if balance.is_zero() => continue,
-                Ok(balance) => balance,
-                Err(err) => {
-                    return match EthcontractErrorType::classify(&err) {
-                        EthcontractErrorType::Node => Err(err.into()),
-                        EthcontractErrorType::Contract => Ok(None),
+        // We use a stream with ready_chunks so that we can start with the addresses of fast
+        // TokenOwnerFinding implementations first without having to wait for slow ones.
+        let stream = self.candidate_owners(token).ready_chunks(BATCH_SIZE);
+        futures::pin_mut!(stream);
+        while let Some(chunk) = stream.next().await {
+            let mut batch = CallBatch::new(self.web3.transport());
+            let futures = chunk
+                .iter()
+                .map(|&address| {
+                    let fut = instance.balance_of(address).batch_call(&mut batch);
+                    async move { (address, fut.await) }
+                })
+                .collect::<Vec<_>>();
+            batch.execute_all(BATCH_SIZE).await;
+            for future in futures {
+                let (address, result) = future.await;
+                let balance = match result {
+                    Ok(balance) if balance.is_zero() => continue,
+                    Ok(balance) => balance,
```

Comment:
nit: How about something like this to keep the logic closer together?
```rust
                match result {
                    Ok(balance) if balance >= min_amount => {
                        return Ok(Some((address, balance)));
                    }
                    Err(err) => {
                        return match EthcontractErrorType::classify(&err) {
                            EthcontractErrorType::Node => Err(err.into()),
                            EthcontractErrorType::Contract => Ok(None),
                        }
                    }
                    _ => continue,
                }
```

Reply:
The main reason it ended up like this is because it was a shorter diff (if you set diff to ignore whitespace changes). The reason it was like this initially is that when one of the match branches feels like the main one to me I like moving it out of the match with a variable to reduce right drift in the text. In this case it hardly matters because now it is only one line and it fits with the other return statements here so I like the suggestion.